### ISSUE TITLE
Upgrade from 4.0.0 to 4.2.1 results in timeout for any fly command

### ISSUE THREAD

**Issue by jschaul** [concourse/concourse#2740]
## Bug Report

After upgrading concourse from 4.0.0 to 4.2.1, and after doing a `fly sync` (which was successful), `fly <any command>` causes a `Get https://<url>/api/v1/info: dial tcp: i/o timeout`, even though the ATC is perfectly reachable 1) via browser and 2) via curl.

Previously posted issue (which received no attention so far - perhaps this is a better place to post it?) is here: https://github.com/concourse/fly/issues/264

## Steps to Reproduce

By downloading previous versions of fly, it is apparent that the problem occurred between `4.0.0` and `4.1.0`:

The following works:

```
curl -Ls https://github.com/concourse/concourse/releases/download/v4.0.0/fly_linux_amd64 > fly400  && chmod +x fly400 && \
./fly400 -t demo login -c 'https://<url>' --verbose
```
(and tells me to sync, see details)
<details>
logging in to team 'main'

2018/09/18 17:52:05 GET /api/v1/info HTTP/1.1
Host: <sanitized>
User-Agent: Go-http-client/1.1
Accept-Encoding: gzip

2018/09/18 17:52:06 HTTP/1.1 200 OK
Content-Length: 43
Cache-Control: no-cache
Connection: keep-alive
Content-Type: application/json
Date: Tue, 18 Sep 2018 15:52:06 GMT
Pragma: no-cache
Server: nginx
X-Concourse-Version: 4.2.1
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Xss-Protection: 1; mode=block

{"version":"4.2.1","worker_version":"2.1"}

WARNING:

fly version (4.0.0) is out of sync with the target (4.2.1). to sync up, run the following:
</details>

whereas this fails with a timeout

```
curl -Ls https://github.com/concourse/concourse/releases/download/v4.1.0/fly_linux_amd64 > fly410  && chmod +x fly410 && \
./fly410 -t demo login -c 'https://<url>' --verbose
```
<details>

```
logging in to team 'main'

2018/09/18 18:11:28 GET /api/v1/info HTTP/1.1
Host: <url>
User-Agent: Go-http-client/1.1
Accept-Encoding: gzip


could not reach the Concourse server called bridge:

    Get https://<url>/api/v1/info: dial tcp: i/o timeout

is the targeted Concourse running? better go catch it lol
```

</details>

## Expected Results

I expect to be able to use `fly <any command>`

## Actual Results

`i/o timeout`

## Additional Context

```sh
curl https://<url>/api/v1/info
```
```json
{"version":"4.2.1","worker_version":"2.1"}
```

Background to our setup:

We front concourse with a nginx server accepting only https traffic, which does SSL termination (There is no http->https redirect, for instance - firewalls prevent any incoming traffic other than on port 443)

sshing onto the ATC server, then running `./fly421 -t demo login -c 'http://127.0.0.1:8080' --verbose` on the ATC server itself works. 

## Version Info

* Concourse version: 4.2.1 (4.1.0 also affected)
* Deployment type (BOSH/Docker/binary): binary
* Infrastructure/IaaS: AWS EC2, ubuntu16 AMI
* Browser (if applicable): n/a
* Did this used to work? yes, any version 2.x, 3.x and including 4.0 works.

**Comment by xtremerui** [comment]
Hi, the only thing could be changed is the Go version. Go 1.11 is shipped with 4.1, and I would image in its net/http package (or the way we used to set it up for concourse client in fly), somehow https is not send to 443 (as your firewall only accept that).

For helping us debug, if you have access to your nginx firewall, could you track the port(or the full context) that the request from fly to /api/v1/info hit?

**Comment by jschaul** [comment]
@ebmeierj no, we're using a single web node. 

Getting the full context for debugging might be a little more involved, but I can try to see if tcpdump can shed any light.

**Comment by NorbertWasilewski** [comment]
I have same issue with `fly` `4.2.1` for linux amd64. Windows executable works well though.

**Comment by dam5s** [comment]
I am seeing the same issue here from my Linux machine. My colleagues that use the MacOSX version of the CLI don't have any issues.

**Comment by dam5s** [comment]
After some investigation, it seems to be some DNS resolution issues with Go 1.11.
If I replace the URL of the concourse server with an IP, it can successfully connect.

It seems that Go 1.11 does not leverage all DNS servers listed in /etc/resolve.conf correctly (if there is more than one).

As a workaround, I configured my DNS connection to use only one DNS server: `1.1.1.1`
And now it works.

**Comment by jschaul** [comment]
Thanks @dam5s !
I can confirm the workaround: If `/etc/resolve.conf` contains two name servers (it does on my local setup usually), it fails. Removing or commenting out name servers such that only one remains in `/etc/resolv.conf` solves the problem.

**Comment by Jeeppler** [comment]
I still experience the timeout issue on Linux with Fly version 4.2.1. However, I only have the time outs while using HTTPS. If I use HTTP everything works fine.

**Comment by sprash98** [comment]
Seeing this on 4.2.2. My installation is on a Google Cloud VM running Ubuntu 18.04. I can login using 'localhost', but supplying the EXTERNAL_URL IP fails,:

` Get http://<EXTERNAL_URL_IP>:8080/api/v1/info: dial tcp <EXTERNAL_URL_IP>:8080: i/o timeout`

But this works:
`fly -t test login -c http://localhost:8080/ -u admin -p <password>
logging in to team \'main\'

target saved`

I have one web node.

**Comment by crsimmons** [comment]
We are running into this on 4.2.2 as well. We have a pipeline that uses [the pipeline resource](https://github.com/concourse/concourse-pipeline-resource) to set itself. The resource is failing to log in. From the worker we can curl `https://<domain>/api/v1/info` successfully but `fly login` fails with:

```
could not reach the Concourse server called a:

    Get https://<domain>/api/v1/info: dial tcp: i/o timeout

is the targeted Concourse running? better go catch it lol
```

Running `fly login -c <private IP of ATC> -k` works though.

Concourse: 4.2.2
IaaS: AWS
Deployment mechanism: BOSH

**Comment by vito** [comment]
To be honest I don't really know what we would investigate here. All I can really suggest is checking your network/firewall configuration. The error is super generic and unlikely to have anything to do with Concourse versions as it can't even connect to the server. If you can reach the server just fine from somewhere else or by using a different address, it's more likely to be a network routing issue.

**Comment by DanielJonesEB** [comment]
Hey @vito - was an error on our part. BOSH Cloud Config was wrong, and we'd erroneously added a superfluous DNS entry (which didn't work). This meant `resolv.conf` had two `nameserver` entries, one of which was fine, and one was garbage.

`curl` didn't seem to have a problem with this; it hanged for a while whilst presumably querying the non-existent DNS server, but eventually failed over to the next one. Using `fly` _did_ result in an eventual query to the extant DNS server (as evidenced by `tcpdump`) but presumably it was just on the cusp of a timeout? I don't know how much DNS resolution is done in the `fly` binary. Anyway, thought this might be of interest to others who get a similar problem, and maybe there's a timeout that could be increased in `fly`?

**Comment by crsimmons** [comment]
Our issue turned out to be a problematic change we had made to our cloud config. This resulted in the `resolv.conf` on our worker instance to be something like

```
cat /etc/resolv.conf
nameserver 10.0.1.2
nameserver 10.0.0.2
search eu-west-2.compute.internal
```

Our mistake was thinking that the reserved DNS IP is always the third IP of the subnet when actually it is the .2 of the VPC super net. It seems that Golang (or maybe fly in particular) is more sensitive to invalid DNS entries than other applications. It was timing out trying to connect over `10.0.1.2` before trying on the valid `10.0.0.2`.

**Comment by vito** [comment]
Thanks for following up!

**Comment by jschaul** [comment]
The issue was related to DNS - `fly` could not deal with multiple DNS entries in the `/etc/resolv.conf`. I had two DNS servers in mine, and taking one out (any one, it wasn't related to a particular DNS server)  was a sucessful workaround.
The latest release of fly and concourse (`5.3.0`) no longer exhibits this problem, so whatever broke from 4.0.0 to 4.2.1 seems to have been fixed by now. So this issue can be closed.

**Comment by xtremerui** [comment]
Thanks for the update. It is probably fixed by bumping GoLang version to go 1.12 in 5.x.
